src/diffCheck/segmentation/DFSegmentation.cc

```diff
@@ -6,95 +6,227 @@
             int nSegments = segmenter.getNumberOfClusters();
 
-            // Get the segments
             for (int indice = 0; indice<nSegments; indice++)
             {
                 std::shared_ptr<geometry::DFPointCloud> segment = std::make_shared<geometry::DFPointCloud>();
                 for (auto pointIndice : clusterToPointMap[indice])
                 {
-                    Eigen::Vector3d point = cilantroPointCloud.points.col(pointIndice).cast<double>();
-                    Eigen::Vector3d normal = cilantroPointCloud.normals.col(pointIndice).cast<double>();
+                    Eigen::Vector3d point = cilantroPointCloud->points.col(pointIndice).cast<double>();
+                    Eigen::Vector3d normal = cilantroPointCloud->normals.col(pointIndice).cast<double>();
                     segment->Points.push_back(point);
                     segment->Normals.push_back(normal);
+                    if (cilantroPointCloud->hasColors())
+                    {
+                        Eigen::Vector3d color = cilantroPointCloud->colors.col(pointIndice).cast<double>();
+                        segment->Colors.push_back(color);
+                    }
                 }
+                if (colorClusters)
+                    segment->ApplyColor(Eigen::Vector3d::Random());
                 segments.push_back(segment);
             }
         }
-        
 
         return segments;
     }
+
+    std::shared_ptr<geometry::DFPointCloud> DFSegmentation::AssociateClusters(
+        std::vector<std::shared_ptr<geometry::DFMesh>> referenceMesh,
+        std::vector<std::shared_ptr<geometry::DFPointCloud>> &clusters,
+        double associationThreshold)
+    {
+        std::shared_ptr<geometry::DFPointCloud> unifiedPointCloud = std::make_shared<geometry::DFPointCloud>();
+        std::vector<std::shared_ptr<geometry::DFPointCloud>> segmentsRemainder;
+
+        // iterate through the mesh faces given as function argument
+        for (std::shared_ptr<diffCheck::geometry::DFMesh> face : referenceMesh)
+        {
+            std::shared_ptr<geometry::DFPointCloud> correspondingSegment;
+            std::shared_ptr<open3d::geometry::TriangleMesh> o3dFace;
+            o3dFace = face->Cvt2O3DTriangleMesh();
+
+            // Getting the center of the mesh face
+            Eigen::Vector3d faceCenter;
+            open3d::geometry::OrientedBoundingBox orientedBoundingBox = o3dFace->GetMinimalOrientedBoundingBox();
+            faceCenter = orientedBoundingBox.GetCenter();
+
+            if (face->NormalsFace.size() == 0)
+            {
+                o3dFace->ComputeTriangleNormals();
+                face->NormalsFace.clear();
+                for (auto normal : o3dFace->triangle_normals_)
+                {
+                    face->NormalsFace.push_back(normal.cast<double>());
+                }
+            }
+            
+            // Getting the normal of the mesh face
+            Eigen::Vector3d faceNormal = face->NormalsFace[0];
```

Comment:
before running this we should make sure that meshes have normals, they might not have it. Let's just print an error message return an empty cloud and raise an error that meshes's should have normals

Reply:
actually the comment was not correctly placed. the check was done a bit above. I have fixed the layout:
```
            // Getting the normal of the mesh face
            if (face->NormalsFace.size() == 0)
            {
                o3dFace->ComputeTriangleNormals();
                face->NormalsFace.clear();
                for (auto normal : o3dFace->triangle_normals_)
                {
                    face->NormalsFace.push_back(normal.cast<double>());
                }
            }
            Eigen::Vector3d faceNormal = face->NormalsFace[0];
```

Reply:
this function to compute normals would be worthy to be integrated in our DFMesh object class

Reply:
done !